### PR TITLE
feat: use watermill as routing framework

### PR DESCRIPTION
**Pull request by smoya**
**Description**

This resolves https://github.com/asyncapi/event-gateway/issues/43.

TL;DR: this sets the base for routing messages by using Watermill.io framework.

Using Watermill.io will reduce the complexity of the code, simplify the routing and message handling and reduce the time spent on all those efforts. Even though more work will be needed around app configuration, things like https://github.com/asyncapi/event-gateway/issues/37 have become now more trivial; adding a new handler to the Watermill router will subscribe to incoming Kafka messages and send them to a Kafka topic. Adding a second handler, those will be consumed by a process that will send them to the websocket clients.

I'm pretty excited for the simplicity of Watermill.io framework and how it simplifies most of the operation.

**Related issue(s)**
https://github.com/asyncapi/event-gateway/issues/43

### CODE REVIEW

File: message/message.go [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
+	watermillmessage "github.com/ThreeDotsLabs/watermill/message"
+)
 
-// Context contains information about the context that surrounds a message.
-type Context struct {
-	Channel string `json:"channel"`
-}
+const (
+	// MetadataChannel is the key used for storing the Channel in the message Metadata.
+	MetadataChannel = "_asyncapi_eg_channel"
```
Comment (magicmatatjahu) [inline message/message.go:10]: `_eg_`? What is this short for?

File: kafka/proxy.go [hunk 2]
```diff
@@ -109,58 +146,70 @@ func (h *produceRequestHandler) Handle(requestKeyVersion *kafkaprotocol.RequestK
...
 
 	return shouldReply, nil
 }
 
-func (h *produceRequestHandler) extractMessages(req sarama.ProduceRequest) []*message.Message {
-	var msgs []*message.Message
-	for topic, r := range req.Records {
-		for _, s := range r {
+func (h *produceRequestHandler) extractMessages(req sarama.ProduceRequest) ([]*watermillmessage.Message, error) {
+	marshaler := watermillkafka.DefaultMarshaler{}
```
Comment (magicmatatjahu) [inline kafka/proxy.go:169]: Does `marshaller` (btw. you missed about second `l`) have to be created for each function call? Maybe it would be better to create singleton outside of function.

File: kafka/proxy.go [hunk 3]
```diff
@@ -109,58 +146,70 @@ func (h *produceRequestHandler) Handle(requestKeyVersion *kafkaprotocol.RequestK
...
 
-	msgs := h.extractMessages(req)
+	msgs, err := h.extractMessages(req)
+	if err != nil {
+		logrus.WithError(err).Error("error extracting messages")
+		return shouldReply, nil
+	}
+
 	if len(msgs) == 0 {
 		logrus.Error("The produce request has no messages")
```
Comment (magicmatatjahu) [inline kafka/proxy.go:156]: `Warn` or `Info` will be better than `Error`

**Review by magicmatatjahu — APPROVED** [review]
Overall good, only two things, but not necessary :)

File: message/message.go (hunk 1 above)
Comment (smoya) [reply message/message.go:10]: The idea is to give a unique-ish key here. `eg` stands for **e**vent **g**ateway.

Comment (magicmatatjahu) [reply message/message.go:10]: At first I thought about the `equivalent` (but where is `q` 😅 ). Ok, thanks for clarification. Maybe add to the comment info about this `eg`?

File: kafka/proxy.go (hunk 3 above)
Comment (smoya) [reply kafka/proxy.go:156]: A Kafka Produce Request **should** contain messages. If there are no messages, something weird happened here. Most probably the proxy is not being able to parse the message correctly?

Perhaps what should be changed is the message to be more explicit: `Unexpected error happened: The produce request has no messages`. 

WDYT?

Comment (magicmatatjahu) [reply kafka/proxy.go:156]: Ok, it can stay :)

File: kafka/proxy.go (hunk 2 above)
Comment (smoya) [reply kafka/proxy.go]: Go uses the term `Marshaler` and `Unmashaler` with just one `l`, both for referring to the interface and for the verb `Marshaling` or `Unmarshaling`. E.g. https://pkg.go.dev/encoding/json#Marshaler

👍  to move it into a var so it is only instantiated once. Btw, this is now done!

File: message/message.go (hunk 1 above)
Comment (smoya) [reply message/message.go:13]: Comment added!

**Review by magicmatatjahu — APPROVED** [review]
LGTM
